roc_jni/src/main/cpp/endpoint.cpp

```diff
@@ -0,0 +1,211 @@
+    roc_endpoint*   endpoint;
+    const char*     uri;
+    jclass          endpointClass;
+    roc_protocol    protocol;
+    int             port;
+    jfieldID        hostAttrId;
+    jfieldID        portAttrId;
+    jfieldID        resourceAttrId;
+    char            buf[128];
+    size_t          bufsz = sizeof(buf);
+
+    if (juri == NULL) {
+        jclass exceptionClass = env->FindClass(ILLEGAL_ARGUMENTS_EXCEPTION);
+        env->ThrowNew(exceptionClass, "Bad uri argument");
+        return;
+    }
+
+    if (roc_endpoint_allocate(&endpoint) != 0 ) {
+        jclass exceptionClass = env->FindClass(EXCEPTION);
+        env->ThrowNew(exceptionClass, "Can't allocate roc_endpoint");
+        return;
+    };
+
+    uri = env->GetStringUTFChars(juri, 0);
+    if (roc_endpoint_set_uri(endpoint, uri) != 0) {
+        env->ReleaseStringUTFChars(juri, uri);
+        roc_endpoint_deallocate(endpoint);
+        jclass exceptionClass = env->FindClass(ILLEGAL_ARGUMENTS_EXCEPTION);
+        env->ThrowNew(exceptionClass, "Bad uri arguments");
+        return;
+    }
+    env->ReleaseStringUTFChars(juri, uri);
+
+    endpointClass = env->FindClass(ENDPOINT_CLASS);
+    assert(endpointClass != NULL);
+
+    roc_endpoint_get_protocol(endpoint, &protocol);
+    endpoint_set_protocol(env, thisObj, protocol);
+
+    roc_endpoint_get_host(endpoint, buf, &bufsz);
+    hostAttrId = env->GetFieldID(endpointClass, "host", "Ljava/lang/String;");
+    assert(hostAttrId != NULL);
+    env->SetObjectField(thisObj, hostAttrId, env->NewStringUTF(buf));
+
+    portAttrId = env->GetFieldID(endpointClass, "port", "I");
+    assert(portAttrId != NULL);
+    if (roc_endpoint_get_port(endpoint, &port) == 0) {
+        env->SetIntField(thisObj, portAttrId, port);
+    } else {
+        env->SetIntField(thisObj, portAttrId, -1);
+    }
+
+    if (roc_endpoint_get_resource(endpoint, buf, &bufsz) == 0) {
+        resourceAttrId = env->GetFieldID(endpointClass, "resource", "Ljava/lang/String;");
+        assert(resourceAttrId != NULL);
+        env->SetObjectField(thisObj, resourceAttrId, env->NewStringUTF(buf));
+    }
+
+    roc_endpoint_deallocate(endpoint);
+}
+
+JNIEXPORT jstring JNICALL Java_org_rocstreaming_roctoolkit_Endpoint_getUri(JNIEnv *env, jobject thisObj) {
+    roc_endpoint*   endpoint;
+    char            buf[128];
+    size_t          bufsz = sizeof(buf);
+
+    if (roc_endpoint_allocate(&endpoint) != 0 ) {
+        jclass exceptionClass = env->FindClass(EXCEPTION);
+        env->ThrowNew(exceptionClass, "Can't allocate roc_endpoint");
+        return NULL;
+    };
+
+    if (endpoint_unmarshal(env, &endpoint, thisObj) != 0) {
+        roc_endpoint_deallocate(endpoint);
+        jclass exceptionClass = env->FindClass(EXCEPTION);
+        env->ThrowNew(exceptionClass, "Can't unmarshal roc_endpoint");
+        return NULL;
+    }
+
+    if (roc_endpoint_get_uri(endpoint, buf, &bufsz) != 0) {
```

Comment:
To avoid limitation of 128 bytes, we should do the following:

* call `roc_endpoint_get_uri(endpoint, NULL, &bufsz)` - it will calculate required buffer size and write it to bufsz
* allocate `buf` of size `bufsz` (e.g. using calloc)
* call `roc_endpoint_get_uri(endpoint, buf, &bufsz)`
* call `env->NewStringUTF(buf)`
* deallocate `buf`
* return java string